glib/src/subclass/object.rs

```diff
@@ -216,20 +282,97 @@ impl<T: ObjectImpl> ObjectImplExt for T {
             }
         }
     }
-
-    fn signal_chain_from_overridden(
+    fn signal_chain_from_overridden<R: TryFromClosureReturnValue>(
         &self,
-        token: &super::SignalClassHandlerToken,
-        values: &[Value],
-    ) -> Option<Value> {
-        unsafe {
-            super::types::signal_chain_from_overridden(
-                self.instance().as_ptr() as *mut _,
-                token,
-                values,
-            )
+        values: &[&dyn ToValue],
+    ) -> R {
+        let mut vs = smallvec::SmallVec::<[Value; 10]>::with_capacity(values.len() + 1);
+        vs.push(self.instance().to_value());
+        vs.extend(values.iter().copied().map(ToValue::to_value));
+
+        R::try_from_closure_return_value(signal_chain_from_overridden(&vs))
+            .expect("Invalid return value")
+    }
+}
+
+// rustdoc-stripper-ignore-next
+/// Calls the original class closure of a signal. This function should only be called from inside
+/// an overridden class closure registered with
+/// [`ObjectClassSubclassExt::override_signal_class_handler`] or
+/// [`ObjectClassSubclassExt::override_signal_class_closure`].
+///
+/// The first value in `values` must be the object currently emitting the signal, and the rest are
+/// the parameters to pass to the parent class closure. Calling this function with an incorrect
+/// number of values or incorrect types for the values will panic. Logic errors may happen if this
+/// function is called more than once inside a class closure or if this function is called outside
+/// a class closure, but it is not undefined behavior.
+#[doc(alias = "g_signal_chain_from_overridden")]
+pub fn signal_chain_from_overridden(values: &[Value]) -> Option<Value> {
+    let instance = values
+        .get(0)
+        .and_then(|v| v.get::<&Object>().ok())
+        .expect("First value in array passed to `signal_chain_from_overridden` must be a GObject");
+    let hint: super::SignalInvocationHint = unsafe {
+        from_glib(gobject_ffi::g_signal_get_invocation_hint(
+            instance.to_glib_none().0,
+        ))
+    };
+    let query = hint.signal_id().query();
```

Comment:
Part of the reason for the token was also to be able to pass around the signal query for type checking purposes. Is this going to be a performance problem if its called for every signal invocation another time?

Reply:
I don't like there being two ways to do this. If we use the token, that should be required for everything. But that means we can't allow closures to be passed directly as override handlers. Instead, if we want to use value unwrapping like in `closure!` we would have to add another macro:

```rust
class.override_signal_class_handler(
    "my-signal",
    glib::override_handler!(
        |token, this: &MyObject, param: &str| -> String {
            // ...
            token.chain(&[&this, &param])
        }
    )
)
```

Reply:
Ok it always uses the token now, callers don't have to pass the original value array but if they do pass it then we can skip the type checks and take the fast path